lib/ide/ast/impl/src/lib.rs

```diff
@@ -584,162 +573,189 @@ pub enum MacroPatternMatchRaw<T> {
+        builder.with_int(*self);
     }
 }
 
-impl<T: HasSpan> HasSpan for Rc<T> {
-    fn span(&self) -> usize {
-        self.deref().span()
+impl Tokenizer for char {
+    fn tokenize(&self, builder:&mut impl TokenBuilder) {
+        builder.with_chr(*self);
     }
 }
 
-impl<T: HasSpan, U: HasSpan> HasSpan for (T,U) {
-    fn span(&self) -> usize {
-        self.0.span() + self.1.span()
+impl Tokenizer for Ast {
+    fn tokenize(&self, builder:&mut impl TokenBuilder) {
+        builder.with_ast(self);
     }
 }
-impl<T: HasSpan, U: HasSpan, V: HasSpan> HasSpan for (T,U,V) {
-    fn span(&self) -> usize {
-        self.0.span() + self.1.span() + self.2.span()
+
+impl<T:Tokenizer> Tokenizer for Option<T> {
+    fn tokenize(&self, builder:&mut impl TokenBuilder) {
+        for t in self { t.tokenize(builder); }
     }
 }
-impl HasSpan for usize {
-    fn span(&self) -> usize {
-        *self
+
+impl<T:Tokenizer> Tokenizer for Vec<T> {
+    fn tokenize(&self, builder:&mut impl TokenBuilder) {
+        for t in self { t.tokenize(builder); }
     }
 }
-impl<T: HasSpan> HasSpan for &T {
-    fn span(&self) -> usize {
-        self.deref().span()
+
+impl<T:Tokenizer> Tokenizer for Rc<T> {
+    fn tokenize(&self, builder:&mut impl TokenBuilder) {
+        self.deref().tokenize(builder);
     }
 }
 
-
-// === HasRepr ===
-
-/// Things that can be asked about their textual representation.
-///
-/// See also `HasSpan`.
-pub trait HasRepr {
-    /// Obtain the text representation for the This type.
-    fn repr(&self) -> String {
-        let mut acc = String::new();
-        self.write_repr(&mut acc);
-        acc
+impl<T:Tokenizer> Tokenizer for &T {
+    fn tokenize(&self, builder:&mut impl TokenBuilder) {
+        self.deref().tokenize(builder);
     }
-
-    fn write_repr(&self, target:&mut String);
 }
 
-impl HasRepr for char {
-    fn write_repr(&self, target:&mut String) {
-        target.push(*self);
+impl<T:Tokenizer,U:Tokenizer> Tokenizer for (T,U) {
+    fn tokenize(&self, builder:&mut impl TokenBuilder) {
+        self.0.tokenize(builder);
+        self.1.tokenize(builder);
     }
 }
-
-impl HasRepr for String {
-    fn write_repr(&self, target:&mut String) {
-        target.push_str(self);
+impl<T:Tokenizer,U:Tokenizer,V:Tokenizer> Tokenizer for (T,U,V) {
+    fn tokenize(&self, builder:&mut impl TokenBuilder) {
+        self.0.tokenize(builder);
+        self.1.tokenize(builder);
+        self.2.tokenize(builder);
     }
 }
 
-impl HasRepr for &str {
-    fn write_repr(&self, target:&mut String) {
-        target.push_str(self);
-    }
+
+// === HasIdMap ===
+
+/// Things that have IdMap.
+trait HasIdMap {
+    /// Extracts IdMap from `self`.
+    fn id_map(&self) -> IdMap;
 }
 
-impl<T: HasRepr> HasRepr for Option<T> {
-    fn write_repr(&self, target:&mut String) {
-        for el in self.iter() {
-            el.write_repr(target)
+struct IdMapBuilder { id_map:IdMap, offset:usize }
```

Comment:
For each defined struct derive `Debug` and `Default`, if possible (and sensible). (and `Copy` if trivial, but that happens rarely).

This comment applies to all struct introduced in this PR.